src/dispatch.rs

```diff
@@ -30,7 +30,29 @@ impl Dispatch {
     // Using helper here to take potentially large stack burden off of call chain to application.
     #[inline(never)]
     fn reply_with_error(&mut self, error: Error) {
-        self.responder.respond(&Err(error)).expect("cant respond");
+        self.reply_or_cancel(&Err(error))
+    }
+
+    fn reply_or_cancel(&mut self, response: &Result<crate::types::Message, Error>) {
+        if self.responder.respond(response).is_ok() {
+            return;
+        }
+
+        let state = self.responder.state();
+
+        match state {
+            interchange::State::Idle
+            | interchange::State::BuildingRequest
+            | interchange::State::BuildingResponse
+            | interchange::State::Requested
+            | interchange::State::Responded => panic!("Unexpected state: {state:?}"),
+            interchange::State::CancelingRequested
+            | interchange::State::CancelingBuildingResponse
+            | interchange::State::Canceled => self
+                .responder
+                .acknowledge_cancel()
+                .expect("failed to cancel"),
```

Comment:
`acknowledge_cancel` requires that the state is `Canceled`.  Why did you also add the other variants?

Reply:
Actually CancelingBuildingResponse might be reachable?

In practice I'd guess no because would be overridden just after by `Canceled` by the usbd-ctaphid crate that has higher priority.